File: examples/settings/swiftui/SettingsTokens.swift
//  SettingsTokens.swift
//  Semantic design-token layer for the settings example.
//
//  This is the ONLY file permitted to hold raw values. Every line carrying a raw
//  number (or a literal color fallback) ends with `// ux:ignore`, so the token
//  linter treats this file as the single source of truth and flags any literal
//  that leaks into product code (the settings UI stays 100% token-driven).
//
//  Colors map to Apple's *semantic* system colors, so light / dark /
//  Increase-Contrast resolve automatically, and a cross-SDK `Color` shim keeps the
//  layer compiling on BOTH the iOS (UIKit) and macOS (AppKit) toolchains — there
//  is no bare `Color(uiColor:)` at top level.
//
//  Spacing lives on a strict 4 / 8 grid (group spacing is a first-class token).
//  Row hit height is >= 48pt (Apple 44 / Material 48). Breakpoints (compact = 600,
//  expanded = 840) drive the responsive single-pane -> two-pane reflow; they mirror
//  the DTCG `breakpoint.*` tokens.

import SwiftUI
#if canImport(UIKit)
import UIKit
#elseif canImport(AppKit)
import AppKit
#endif

enum SettingsTokens {

    // MARK: Color — semantic roles (auto light / dark / high-contrast)
    // Sourced from platform system colors so light/dark/Increase-Contrast resolve
    // automatically. Cross-platform shim lives in the `Color` extension below.
    //
    // Contrast (verified against WCAG 2.2 §1.4.3, both themes):
    //   • on-surface = label on grouped background            ≈ 15:1  (≥ 4.5:1)
    //   • on-surface-variant = secondaryLabel on background   ≈ 7:1   (≥ 4.5:1)
    //   • status-error (destructive label) on surface         ≥ 4.5:1
    //   • outline-variant divider / switch track vs surface   ≥ 3:1   (§1.4.11)
    // The destructive label is ALWAYS paired with a confirm dialog and an icon,
    // never color alone (COL-003 / A11Y-012).

    static let surface          = Color.uxSurface
    static let surfaceContainer = Color.uxSurfaceContainer
    static let onSurface        = Color.uxOnSurface
    static let onSurfaceVariant = Color.uxOnSurfaceVariant
    static let outlineVariant   = Color.uxOutlineVariant

    static let actionPrimary    = Color.accentColor
    static let onActionPrimary  = Color.white
    static let statusError      = Color.uxStatusError
    static let statusSuccess    = Color.uxStatusSuccess
    static let statusWarning    = Color.uxStatusWarning

    /// Neutral skeleton fill for shape-matched loading placeholders (synced values).
    static let skeleton = Color.uxSkeleton

    // MARK: Spacing — 4 / 8 grid

    static let s1: CGFloat = 4    // ux:ignore
    static let s2: CGFloat = 8    // ux:ignore
    static let s3: CGFloat = 12   // ux:ignore
    static let s4: CGFloat = 16   // ux:ignore
    static let s5: CGFloat = 20   // ux:ignore
    static let s6: CGFloat = 24   // ux:ignore
    static let s8: CGFloat = 32   // ux:ignore

    /// Vertical spacing BETWEEN grouped sections (SPC-006). A first-class token so
    /// group gutters stay in lock-step with the grouped inset style.
    static let groupSpacing: CGFloat = 24   // ux:ignore

    // MARK: Radius (concentric grouped-row corners on iOS, SHP-003)

    static let rowRadius: CGFloat     = 12   // ux:ignore
    static let controlRadius: CGFloat = 8    // ux:ignore
    static let bannerRadius: CGFloat  = 12   // ux:ignore

    // MARK: Size / target — >= 48pt hit areas (Apple 44 / Material 48)

    static let targetMin: CGFloat    = 48    // ux:ignore
    /// Minimum tappable row height. >= 48 satisfies Apple 44pt AND Material 48dp.
    static let rowMinHeight: CGFloat = 48    // ux:ignore
    static let hairline: CGFloat     = 1     // ux:ignore
    static let iconColumn: CGFloat   = 28    // ux:ignore

    /// Sidebar (leading pane) ideal width in the two-pane layout.
    static let sidebarMinWidth: CGFloat   = 240   // ux:ignore
    static let sidebarIdealWidth: CGFloat = 300   // ux:ignore

    /// Cap the content measure so a settings column never stretches edge-to-edge on
    /// a very wide display (GRD-005 / SPC-018).
    static let contentMaxWidth: CGFloat = 680   // ux:ignore

    /// Skeleton shape metrics (non-text placeholders → fixed frames are fine).
    static let skeletonValueWidth: CGFloat = 72   // ux:ignore
    static let skeletonLineHeight: CGFloat = 14   // ux:ignore

    // MARK: Breakpoints — window-size-class tokens (GRD-004)
    // Compact < 600 (single scrolling list; a group pushes a sub-page) ·
    // Expanded >= 840 (two-pane: group list + selected group's settings).

    static let breakpointCompact: CGFloat  = 600   // ux:ignore
    static let breakpointExpanded: CGFloat = 840   // ux:ignore

    // MARK: Typography — text-style roles that scale with Dynamic Type

    static let screenTitleFont  = Font.title2
    static let sectionHeaderFont = Font.footnote
    static let rowTitleFont     = Font.body
    static let rowValueFont     = Font.body
    static let rowSubtitleFont  = Font.footnote
    static let footerFont       = Font.footnote
    static let bannerFont       = Font.subheadline

    // MARK: Motion — Reduce-Motion decision lives with the token (MOT-004)
    // Only opacity / position transitions are used, and each collapses to an
    // ~instant change under Reduce Motion.

    private static let instantDuration: Double = 0.001   // ux:ignore
    private static let revealDuration: Double  = 0.20    // ux:ignore  ≤ 200ms

    /// Banner / value cross-fade. ~instant under Reduce Motion.
    static func reveal(reduceMotion: Bool) -> Animation {
        reduceMotion
            ? .linear(duration: instantDuration)
            : .easeInOut(duration: revealDuration)
    }

    // MARK: Demo timing (stand-ins for a real fetch / save round-trip)

    static let syncDelay: Duration = .milliseconds(700)   // ux:ignore
    static let saveDelay: Duration = .milliseconds(450)   // ux:ignore
    static let toastLinger: Duration = .seconds(2)        // ux:ignore
}

// MARK: - Cross-platform system colors
// Resolves Apple semantic colors on iOS (UIKit) and macOS (AppKit) so the token
// layer compiles on both SDKs. In a shipping iOS app these map to asset-catalog
// Color sets generated from DTCG tokens.
private extension Color {
    static var uxSurface: Color {
        #if canImport(UIKit)
        Color(uiColor: .systemGroupedBackground)
        #elseif canImport(AppKit)
        Color(nsColor: .windowBackgroundColor)
        #else
        Color.white
        #endif
    }
    static var uxSurfaceContainer: Color {
        #if canImport(UIKit)
        Color(uiColor: .secondarySystemGroupedBackground)
        #elseif canImport(AppKit)
        Color(nsColor: .underPageBackgroundColor)
        #else
        Color.gray.opacity(0.12)   // ux:ignore
        #endif
    }
    static var uxOnSurface: Color {
        #if canImport(UIKit)
        Color(uiColor: .label)
        #elseif canImport(AppKit)
        Color(nsColor: .labelColor)
        #else
        Color.primary
        #endif
    }
    static var uxOnSurfaceVariant: Color {
        #if canImport(UIKit)
        Color(uiColor: .secondaryLabel)
        #elseif canImport(AppKit)
        Color(nsColor: .secondaryLabelColor)
        #else
        Color.secondary
        #endif
    }
    static var uxOutlineVariant: Color {
        #if canImport(UIKit)
        Color(uiColor: .separator)
        #elseif canImport(AppKit)
        Color(nsColor: .separatorColor)
        #else
        Color.gray
        #endif
    }
    static var uxStatusError: Color {
        #if canImport(UIKit)
        Color(uiColor: .systemRed)
        #elseif canImport(AppKit)
        Color(nsColor: .systemRed)
        #else
        Color.red
        #endif
    }
    static var uxStatusSuccess: Color {
        #if canImport(UIKit)
        Color(uiColor: .systemGreen)
        #elseif canImport(AppKit)
        Color(nsColor: .systemGreen)
        #else
        Color.green
        #endif
    }
    static var uxStatusWarning: Color {
        #if canImport(UIKit)
        Color(uiColor: .systemOrange)
        #elseif canImport(AppKit)
        Color(nsColor: .systemOrange)
        #else
        Color.orange
        #endif
    }
    static var uxSkeleton: Color {
        #if canImport(UIKit)
        Color(uiColor: .tertiarySystemFill)
        #elseif canImport(AppKit)
        Color(nsColor: .quaternaryLabelColor)
        #else
        Color.gray.opacity(0.2)   // ux:ignore
        #endif
    }
}

File: examples/settings/jetpack-compose/SettingsTokens.kt
package ux.examples.settings

/**
 * SettingsTokens — the semantic token layer for the settings example.
 *
 * This is the ONLY file in this example allowed to contain raw values (hex colors,
 * dp spacing / size / radius, the breakpoint dp values, and millis). Every such line is
 * annotated `// ux:ignore` so the token_lint validator treats this file as the single
 * source of literals. The settings composable consumes ONLY these tokens plus
 * `MaterialTheme.colorScheme` / `.typography` / `.shapes` roles — never a literal.
 *
 * Layering (DTCG -> Compose): color / typography / shape flow through
 * `MaterialExpressiveTheme` (so dark mode + Material You resolve through role slots);
 * spacing / size / radius / breakpoints — which M3 has no theme slot for — are carried as
 * plain token objects (`Space`, `Size`, `Radius`, `Breakpoints`). All the roles the settings
 * list needs map cleanly onto the M3 ColorScheme slots, so no CompositionLocal is required:
 *   • row background / grouped card    -> surface / surfaceContainer
 *   • row divider + group inset        -> outlineVariant
 *   • row label / value / header       -> onSurface / onSurfaceVariant
 *   • switch on-track / thumb          -> primary / onPrimary
 *   • destructive label (sign out …)   -> error / onError
 */

import android.provider.Settings
import androidx.compose.foundation.isSystemInDarkTheme
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.ColorScheme
import androidx.compose.material3.ExperimentalMaterial3ExpressiveApi
import androidx.compose.material3.MaterialExpressiveTheme
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Shapes
import androidx.compose.material3.darkColorScheme
import androidx.compose.material3.lightColorScheme
import androidx.compose.runtime.Composable
import androidx.compose.runtime.ReadOnlyComposable
import androidx.compose.runtime.remember
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Shape
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp

// ─────────────────────────────────────────────────────────────────────────────
// SPACING — DTCG space.* scale, snapped to the 4/8 grid (SPC-*). `rowInset` is the
// 16dp leading keyline (space.4) every row and divider aligns to; `group` is the 24dp
// gap between grouped sections (space.6 per the spec's token table).
// ─────────────────────────────────────────────────────────────────────────────
object Space {
    val zero: Dp = 0.dp   // ux:ignore  (animation rest / no-gap)
    val xs: Dp = 4.dp     // ux:ignore
    val sm: Dp = 8.dp     // ux:ignore
    val md: Dp = 16.dp    // ux:ignore
    val lg: Dp = 24.dp    // ux:ignore
    val xl: Dp = 32.dp    // ux:ignore

    /** Row leading keyline == divider inset (space.4). */
    val rowInset: Dp = 16.dp // ux:ignore

    /** Spacing between grouped sections (space.6). */
    val group: Dp = 24.dp    // ux:ignore
}

// ─────────────────────────────────────────────────────────────────────────────
// SIZE — target minimums + glyph / pane sizes (A11Y-*, ICN-*). `rowMinHeight` is the
// Material 48dp minimum every interactive row / switch / destructive row clears.
// ─────────────────────────────────────────────────────────────────────────────
object Size {
    /** Material minimum interactive target — every row clears it (also the iOS 44pt floor). */
    val rowMinHeight: Dp = 48.dp     // ux:ignore
    val icon: Dp = 24.dp             // ux:ignore
    val emptyIcon: Dp = 40.dp        // ux:ignore  (zero-results illustration glyph)

    // Shape-matched skeleton block heights (non-text placeholder boxes for synced values).
    val skeletonLabel: Dp = 16.dp    // ux:ignore
    val skeletonValue: Dp = 12.dp    // ux:ignore

    /** Leading category pane width in the expanded (≥840dp) two-pane layout. */
    val railWidth: Dp = 320.dp       // ux:ignore

    /** Max content measure so the list never stretches edge-to-edge on very wide screens. */
    val maxContentWidth: Dp = 640.dp // ux:ignore
}

// ─────────────────────────────────────────────────────────────────────────────
// RADIUS / SHAPES — grouped-card radius (lg) + field radius (md); the fully-rounded
// pill is kept for completeness (SHP-*).
// ─────────────────────────────────────────────────────────────────────────────
object Radius {
    val sm: Dp = 8.dp      // ux:ignore
    val md: Dp = 12.dp     // ux:ignore
    val lg: Dp = 16.dp     // ux:ignore  (grouped inset cards)
    val pill: Dp = 9999.dp // ux:ignore  (fully-rounded control, if needed)
}

val SettingsShapes = Shapes(
    extraSmall = RoundedCornerShape(Radius.sm),
    small = RoundedCornerShape(Radius.sm),
    medium = RoundedCornerShape(Radius.md),
    large = RoundedCornerShape(Radius.lg),
    extraLarge = RoundedCornerShape(Radius.lg),
)

/** Fully-rounded shape, available for pill-styled controls. */
val PillShape: Shape = RoundedCornerShape(Radius.pill)

// ─────────────────────────────────────────────────────────────────────────────
// BREAKPOINTS — M3 window-size-class widths (GRD-004). Compact < 600dp is a single
// scrolling list that pushes a sub-page; Expanded ≥ 840dp is a two-pane list-detail.
// The 600–839dp medium band stays single-pane. Same values M3's WindowSizeClass uses.
// ─────────────────────────────────────────────────────────────────────────────
object Breakpoints {
    val compact: Dp = 600.dp   // ux:ignore
    val expanded: Dp = 840.dp  // ux:ignore
}

// ─────────────────────────────────────────────────────────────────────────────
// MOTION — durations for the (alpha-only) banner transitions; reduce-motion collapses
// these to instant (see rememberReduceMotion). Only opacity is ever animated. The "*Ms"
// values are demo-only stand-ins for real save / sync latency (MOT-*).
// ─────────────────────────────────────────────────────────────────────────────
object Motion {
    const val shortMillis = 150       // ux:ignore  (banner fade ≤200ms, no bounce)
    const val simulatedSaveMs = 900L  // ux:ignore  (demo: server save round-trip)
    const val simulatedSyncMs = 1200L // ux:ignore  (demo: initial synced-value fetch)
}

// ─────────────────────────────────────────────────────────────────────────────
// COLOR — brand fallback palette. DTCG semantic roles are mapped onto M3 ColorScheme
// slots below; MaterialExpressiveTheme resolves light / dark (and Material You can be
// layered on Android 12+). Contrast pairs meet WCAG 2.2 (COL-*, DRK-*, A11Y-*).
// ─────────────────────────────────────────────────────────────────────────────
private object Palette {
    // Light
    val primaryLight = Color(0xFF4C5BC4)              // ux:ignore
    val onPrimaryLight = Color(0xFFFFFFFF)            // ux:ignore
    val primaryContainerLight = Color(0xFFE0E1FF)     // ux:ignore
    val onPrimaryContainerLight = Color(0xFF03105B)   // ux:ignore
    val secondaryLight = Color(0xFF5A5D72)            // ux:ignore
    val onSecondaryLight = Color(0xFFFFFFFF)          // ux:ignore
    val secondaryContainerLight = Color(0xFFDFE1F9)   // ux:ignore
    val onSecondaryContainerLight = Color(0xFF171B2C) // ux:ignore
    val surfaceLight = Color(0xFFFBF8FF)              // ux:ignore
    val onSurfaceLight = Color(0xFF1A1B21)            // ux:ignore
    val surfaceVariantLight = Color(0xFFE3E1EC)       // ux:ignore
    val onSurfaceVariantLight = Color(0xFF45464F)     // ux:ignore
    val surfaceContainerLight = Color(0xFFEFEDF7)     // ux:ignore
    val outlineLight = Color(0xFF767680)              // ux:ignore
    val outlineVariantLight = Color(0xFFC6C5D0)       // ux:ignore
    val errorLight = Color(0xFFB3261E)                // ux:ignore
    val onErrorLight = Color(0xFFFFFFFF)              // ux:ignore
    val errorContainerLight = Color(0xFFF9DEDC)       // ux:ignore
    val onErrorContainerLight = Color(0xFF410E0B)     // ux:ignore

    // Dark
    val primaryDark = Color(0xFFBEC2FF)               // ux:ignore
    val onPrimaryDark = Color(0xFF1B2678)             // ux:ignore
    val primaryContainerDark = Color(0xFF333EA0)      // ux:ignore
    val onPrimaryContainerDark = Color(0xFFE0E1FF)    // ux:ignore
    val secondaryDark = Color(0xFFC3C5DD)             // ux:ignore
    val onSecondaryDark = Color(0xFF2C2F42)           // ux:ignore
    val secondaryContainerDark = Color(0xFF424659)    // ux:ignore
    val onSecondaryContainerDark = Color(0xFFDFE1F9)  // ux:ignore
    val surfaceDark = Color(0xFF121318)               // ux:ignore
    val onSurfaceDark = Color(0xFFE3E1E9)             // ux:ignore
    val surfaceVariantDark = Color(0xFF45464F)        // ux:ignore
    val onSurfaceVariantDark = Color(0xFFC6C5D0)      // ux:ignore
    val surfaceContainerDark = Color(0xFF1E1F25)      // ux:ignore
    val outlineDark = Color(0xFF90909A)               // ux:ignore
    val outlineVariantDark = Color(0xFF45464F)        // ux:ignore
    val errorDark = Color(0xFFF2B8B5)                 // ux:ignore
    val onErrorDark = Color(0xFF601410)               // ux:ignore
    val errorContainerDark = Color(0xFF8C1D18)        // ux:ignore
    val onErrorContainerDark = Color(0xFFF9DEDC)      // ux:ignore
}

private val SettingsLightColors: ColorScheme = lightColorScheme(
    primary = Palette.primaryLight,
    onPrimary = Palette.onPrimaryLight,
    primaryContainer = Palette.primaryContainerLight,
    onPrimaryContainer = Palette.onPrimaryContainerLight,
    secondary = Palette.secondaryLight,
    onSecondary = Palette.onSecondaryLight,
    secondaryContainer = Palette.secondaryContainerLight,
    onSecondaryContainer = Palette.onSecondaryContainerLight,
    background = Palette.surfaceLight,
    onBackground = Palette.onSurfaceLight,
    surface = Palette.surfaceLight,
    onSurface = Palette.onSurfaceLight,
    surfaceVariant = Palette.surfaceVariantLight,
    onSurfaceVariant = Palette.onSurfaceVariantLight,
    surfaceContainer = Palette.surfaceContainerLight,
    outline = Palette.outlineLight,
    outlineVariant = Palette.outlineVariantLight,
    error = Palette.errorLight,
    onError = Palette.onErrorLight,
    errorContainer = Palette.errorContainerLight,
    onErrorContainer = Palette.onErrorContainerLight,
)

private val SettingsDarkColors: ColorScheme = darkColorScheme(
    primary = Palette.primaryDark,
    onPrimary = Palette.onPrimaryDark,
    primaryContainer = Palette.primaryContainerDark,
    onPrimaryContainer = Palette.onPrimaryContainerDark,
    secondary = Palette.secondaryDark,
    onSecondary = Palette.onSecondaryDark,
    secondaryContainer = Palette.secondaryContainerDark,
    onSecondaryContainer = Palette.onSecondaryContainerDark,
    background = Palette.surfaceDark,
    onBackground = Palette.onSurfaceDark,
    surface = Palette.surfaceDark,
    onSurface = Palette.onSurfaceDark,
    surfaceVariant = Palette.surfaceVariantDark,
    onSurfaceVariant = Palette.onSurfaceVariantDark,
    surfaceContainer = Palette.surfaceContainerDark,
    outline = Palette.outlineDark,
    outlineVariant = Palette.outlineVariantDark,
    error = Palette.errorDark,
    onError = Palette.onErrorDark,
    errorContainer = Palette.errorContainerDark,
    onErrorContainer = Palette.onErrorContainerDark,
)

// ─────────────────────────────────────────────────────────────────────────────
// TYPOGRAPHY — semantic text roles resolve to MaterialTheme.typography, so every label
// scales with the user's font size / Dynamic Type (TYP-*, A11Y-*). No raw sp. Maps the
// spec's type.body.md (row label / value) and type.label.md (group header) roles.
// ─────────────────────────────────────────────────────────────────────────────
object SettingsType {
    val screenTitle: TextStyle @Composable @ReadOnlyComposable get() = MaterialTheme.typography.headlineSmall
    val groupHeader: TextStyle @Composable @ReadOnlyComposable get() = MaterialTheme.typography.titleSmall
    val rowLabel: TextStyle @Composable @ReadOnlyComposable get() = MaterialTheme.typography.bodyLarge
    val rowValue: TextStyle @Composable @ReadOnlyComposable get() = MaterialTheme.typography.bodyMedium
    val emptyTitle: TextStyle @Composable @ReadOnlyComposable get() = MaterialTheme.typography.titleMedium
    val body: TextStyle @Composable @ReadOnlyComposable get() = MaterialTheme.typography.bodyMedium
}

// ─────────────────────────────────────────────────────────────────────────────
// THEME — one provider at the root. Dark via isSystemInDarkTheme(); the Expressive
// theme carries the MotionScheme so reduce-motion is honored centrally (DRK-*, MOT-*).
// The app can override the system value with an explicit light/dark choice (see the
// theme selector in SettingsScreen) by passing `darkTheme`.
// ─────────────────────────────────────────────────────────────────────────────
@OptIn(ExperimentalMaterial3ExpressiveApi::class)
@Composable
fun SettingsTheme(
    darkTheme: Boolean = isSystemInDarkTheme(),
    content: @Composable () -> Unit,
) {
    val colorScheme = if (darkTheme) SettingsDarkColors else SettingsLightColors
    MaterialExpressiveTheme(
        colorScheme = colorScheme,
        shapes = SettingsShapes,
        content = content,
    )
}

// ─────────────────────────────────────────────────────────────────────────────
// REDUCE MOTION — read the system animator scale once; when it's 0 ("Remove animations"
// in Accessibility settings) all transitions collapse to instant (MOT-*, A11Y-*).
// ─────────────────────────────────────────────────────────────────────────────
@Composable
fun rememberReduceMotion(): Boolean {
    val resolver = LocalContext.current.contentResolver
    return remember(resolver) {
        Settings.Global.getFloat(resolver, Settings.Global.ANIMATOR_DURATION_SCALE, 1f) == 0f // ux:ignore
    }
}
